File: src/main/java/de/fluxparticle/animation/binding/BindingObservableFunction.java
package de.fluxparticle.animation.binding;

import javafx.beans.InvalidationListener;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.ReadOnlyObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;

import java.util.function.Function;

public class BindingObservableFunction<T, R extends ObservableValue<V>, V> extends ReadOnlyObjectProperty<V> {

	private final ObservableValue<R> binding;

	private final ObjectProperty<V> value;

	public BindingObservableFunction(ReadOnlyObjectProperty<T> observable, Function<T, R> func, String name) {
		this.binding = new BindingFunction<>(observable, func);

		value = new SimpleObjectProperty<>(observable, observable.getName() + '.' + name);
		value.bind(binding.getValue());
		binding.addListener((o, oldBinding, newBinding) -> {
			value.bind(newBinding);
		});
	}

	@Override
	public void addListener(InvalidationListener listener) {
		value.addListener(listener);
	}

	@Override
	public void removeListener(InvalidationListener listener) {
		value.removeListener(listener);
	}

	@Override
	public void addListener(ChangeListener<? super V> listener) {
		value.addListener(listener);
	}

	@Override
	public void removeListener(ChangeListener<? super V> listener) {
		value.removeListener(listener);
	}

	@Override
	public V get() {
		return value.get();
	}

	@Override
	public Object getBean() {
		return value.getBean();
	}

	@Override
	public String getName() {
		return value.getName();
	}

}

File: src/main/kotlin/de/fluxparticle/animation/binding/BindingObservableFunction.kt
package de.fluxparticle.animation.binding

/*
class BindingObservableFunction<T, R : ObservableValue<V>, V>(observable: ReadOnlyObjectProperty<T>, func: (T) -> R, name: String) : ReadOnlyObjectProperty<V>() {

    private val binding: ObservableValue<R>

    private val value: ObjectProperty<V>

    init {
        this.binding = BindingFunction(observable, func)

        value = SimpleObjectProperty(observable, observable.name + '.' + name)
        value.bind(binding.value)
        binding.addListener { o, oldBinding, newBinding -> value.bind(newBinding) }
    }

    override fun addListener(listener: InvalidationListener) {
        value.addListener(listener)
    }

    override fun removeListener(listener: InvalidationListener) {
        value.removeListener(listener)
    }

    override fun addListener(listener: ChangeListener<in V>) {
        value.addListener(listener)
    }

    override fun removeListener(listener: ChangeListener<in V>) {
        value.removeListener(listener)
    }

    override fun get(): V {
        return value.get()
    }

    override fun getBean(): Any {
        return value.bean
    }

    override fun getName(): String {
        return value.name
    }

}
*/
